src/Microsoft.CodeQuality.Analyzers/CSharp/QualityGuidelines/CSharpRemoveEmptyFinalizersAnalyzer.cs

```diff
@@ -13,39 +14,79 @@ internal class CSharpRemoveEmptyFinalizersAnalyzer : AbstractRemoveEmptyFinalize
     {
         protected override bool IsEmptyFinalizer(SyntaxNode methodBody, CodeBlockAnalysisContext analysisContext)
         {
-            var destructorDeclaration = (DestructorDeclarationSyntax)methodBody;
+            // TODO: Switch the analyzer to be IOperation based to avoid handling different syntactic constructs.
 
-            if ((destructorDeclaration?.Body?.Statements.Count ?? 0) == 0)
+            var destructorDeclaration = (DestructorDeclarationSyntax)methodBody;
+            if (destructorDeclaration.Body != null)
             {
-                return true;
+                // Bail out for error case where both expression and block body are non-null.
```

Comment:
❔ Would this violate an invariant of the data model? i.e. can we replace this check with a simple `Debug.Assert` call?

Reply:
This is actually a valid possibility for error cases - in fact our IOperation tree also exposes 2 operation blocks for this case and we have IOperation unit tests to verify there are 2 operation blocks. Added unit test `CA1821CSharpRemoveEmptyFinalizers_ErrorCodeWithBothBlockAndExpressionBody`